File: packages/python-sdk/src/superserve/_telemetry.py
"""Lightweight, fire-and-forget telemetry for the Superserve Python SDK."""

from __future__ import annotations

import importlib.metadata
import os
import platform
import threading
import uuid
from typing import Any

_POSTHOG_KEY = "phc_gjpDKKKQJAnkxkqLrPGrAhoariKsaHNuTpI5rVhkYre"
_POSTHOG_URL = "https://us.i.posthog.com/capture/"

_distinct_id = str(uuid.uuid4())
_disabled = (
    os.environ.get("DO_NOT_TRACK") == "1"
    or os.environ.get("SUPERSERVE_DO_NOT_TRACK") == "1"
    or os.environ.get("SUPERSERVE_TELEMETRY") == "0"
)


def _get_version() -> str:
    try:
        return importlib.metadata.version("superserve")
    except Exception:
        return "unknown"


def _send(event: str, properties: dict[str, Any]) -> None:
    try:
        import httpx

        httpx.post(
            _POSTHOG_URL,
            json={
                "api_key": _POSTHOG_KEY,
                "event": event,
                "distinct_id": _distinct_id,
                "properties": {
                    "sdk": "python",
                    "sdk_version": _get_version(),
                    "python_version": platform.python_version(),
                    **properties,
                },
            },
            timeout=5.0,
        )
    except Exception:
        pass


def track_event(event: str, properties: dict[str, Any] | None = None) -> None:
    """Fire-and-forget telemetry event. Never throws, never blocks."""
    if _disabled:
        return
    threading.Thread(
        target=_send,
        args=(event, properties or {}),
        daemon=True,
    ).start()

File: packages/sdk/src/telemetry.ts
/* eslint-disable no-var */
declare const __SDK_VERSION__: string
declare var process: { env: Record<string, string | undefined> } | undefined
declare var Bun: unknown

const POSTHOG_KEY = "phc_gjpDKKKQJAnkxkqLrPGrAhoariKsaHNuTpI5rVhkYre"
const POSTHOG_URL = "https://us.i.posthog.com/capture/"

let _disabled: boolean | null = null
const _distinctId =
  typeof crypto !== "undefined" && crypto.randomUUID
    ? crypto.randomUUID()
    : Math.random().toString(36).slice(2)

function isDisabled(): boolean {
  if (_disabled === null) {
    const env =
      typeof process !== "undefined" && process?.env
        ? process.env
        : ({} as Record<string, string | undefined>)
    _disabled =
      env.DO_NOT_TRACK === "1" ||
      env.SUPERSERVE_DO_NOT_TRACK === "1" ||
      env.SUPERSERVE_TELEMETRY === "0"
  }
  return _disabled
}

function getRuntime(): string {
  try {
    if (typeof Bun !== "undefined") return "bun"
  } catch {}
  return "node"
}

/**
 * Fire-and-forget telemetry event. Never throws, never blocks.
 * Respects DO_NOT_TRACK and SUPERSERVE_DO_NOT_TRACK env vars.
 */
export function trackEvent(
  event: string,
  properties?: Record<string, unknown>,
): void {
  if (isDisabled()) return
  try {
    const version =
      typeof __SDK_VERSION__ !== "undefined" ? __SDK_VERSION__ : "unknown"
    fetch(POSTHOG_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        api_key: POSTHOG_KEY,
        event,
        distinct_id: _distinctId,
        properties: {
          sdk: "typescript",
          sdk_version: version,
          runtime: getRuntime(),
          ...properties,
        },
      }),
    }).catch(() => {})
  } catch {
    // Never throw from telemetry
  }
}
